pkg/config/configuration.go

```diff
@@ -0,0 +1,186 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"gopkg.in/yaml.v3"
+)
+
+// File is an on-disk configuration of a Config.
+type Configuration struct {
+	Registry Registry `mapstructure:"registry"`
+	TokenCmd string   `mapstructure:"token_cmd" yaml:"token_cmd"` // run in shell to generate token
+}
+
+type Registry struct {
+	Address  string `mapstructure:"address" yaml:"address"`   // service address
+	Insecure bool   `mapstructure:"insecure" yaml:"insecure"` // if true, connect over HTTP
+	Token    string `mapstructure:"token" yaml:"-"`           // generated from TokenCmd
+}
+
+// Write stores the Configuration in the passed file name.
+func (c Configuration) Write(name string) error {
+	if err := ValidateName(name); err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(Directory, os.FileMode(0755)); err != nil { // rwx,rx,rx
+		return err
+	}
+	path := filepath.Join(Directory, name)
+	out, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(0644)) // rw,r,r
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	enc := yaml.NewEncoder(out)
+	return enc.Encode(c)
+}
+
+// Validate returns an error if Config is invalid.
+func (c Configuration) Validate() error {
+	if c.Registry.Address == "" {
+		return ValidationError{
+			"registry.address", "required",
+		}
+	}
+	return nil
+}
+
+// Properties returns a sorted list of all valid property names.
+func (c Configuration) Properties() []string {
+	props := properties(c, "")
+	sort.Strings(props)
+	return props
+}
+
+// FlatMap returns a Configuration as a flat Map.
+func (c Configuration) FlatMap() (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+	flat := map[string]interface{}{}
+	err := mapstructure.Decode(c, &m)
+	flattenMap(m, flat, "")
+	return flat, err
+}
+
+// Set sets a property from a qualified name.
+func (c *Configuration) Set(k string, v interface{}) error {
+	return c.FromMap(map[string]interface{}{
+		k: v,
+	})
+}
+
+// Unset removed a property by qualified name.
+func (c *Configuration) Unset(k string) error {
+	return c.FromMap(map[string]interface{}{
+		k: "",
+	})
+}
+
+// FromMap populates a Configuration from a flat Map.
+// Existing values are overridden.
+func (c *Configuration) FromMap(m map[string]interface{}) error {
+	m = unflattenMap(m)
+
+	config := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           c,
+	}
+
+	decoder, err := mapstructure.NewDecoder(config)
+	if err != nil {
+		return err
+	}
+
+	return decoder.Decode(m)
+}
+
+// Resolve derived values (eg. Registry.Token from Registry.TokenCmd)
+func (c *Configuration) Resolve() error {
```

Comment:
How frequently is this run? 
1. before every token usage? 
2. once for each run of the registry tool?
3. when tokens expire?

Reply:
Currently, this would only happen on load - ie. #2.